### PR TITLE
Extend CSR constructor to support batched indices and values

### PR DESCRIPTION
**Pull request by IvanYashchuk**
This is the first portion of changes required to enable Batched CSR format described in https://github.com/pytorch/pytorch/issues/60854#batched-CSR-computation.

Currently, only the same batch shape for indices and values is allowed. In the future, we could enable "broadcasting" of indices and batched values, as done in xFormers (https://github.com/facebookresearch/xformers/blob/dd96b8d8beda5308fb433c1ef3ff04b7f178c263/xformers/components/attention/_sputnik_sparse.py#L441).

This PR adds possibility to construct a batched CSR matrix with `torch.sparse_csr_tensor` and this batched CSR can be converted to a dense tensor with a `.to_dense()` call.

### CODE REVIEW

File: aten/src/ATen/native/sparse/SparseCsrTensor.cpp [hunk 1]
```diff
@@ -213,13 +242,10 @@ Tensor sparse_csr_tensor(
...
-  if (col_indices.numel() > 0) {
-    AT_DISPATCH_INDEX_TYPES(col_indices.scalar_type(), "csr_construct_check", [&] {
-      size[0] = crow_indices.numel() - 1;
-      size[1] = col_indices.max().item<index_t>() + 1;
-    });
-  }
+  // std::array<int64_t, 2> size = {0, 0};
+  auto size = DimVector(IntArrayRef(col_indices.sizes().data(), col_indices.dim() - 1));
+  size.push_back(crow_indices.size(-1) - 1);
+  size.push_back(col_indices.max().item<int64_t>() + 1);
```
Comment (cpuhrsch) [inline aten/src/ATen/native/sparse/SparseCsrTensor.cpp:248]: col_indices are always guaranteed to be int64_t now?

Comment (IvanYashchuk) [reply aten/src/ATen/native/sparse/SparseCsrTensor.cpp:248]: No, but here `.max().item()` gives a Scalar and `.item<int64_t>` casts the Scalar to `int64_t`.
https://github.com/pytorch/pytorch/blob/45da320092a6e23774a9cca5d7aaeb77abce2f99/aten/src/ATen/templates/TensorMethods.cpp#L25

File: test/test_sparse_csr.py [hunk 2]
```diff
@@ -218,23 +247,22 @@ def test_empty(self, device, dtype):
...
-                                    dtype=dtype,
-                                    device=device)
-        y = x.clone()
-
-        self.assertEqual(x, y)
+        from operator import mul
+        from functools import reduce
+        for batch_shape in ((), (2,), (2, 3)):
+            prod = reduce(mul, batch_shape, 1)
+            crow_indices = torch.tensor([0, 2, 4], device=device).repeat(prod, 1).reshape(*batch_shape, -1)
```
Comment (cpuhrsch) [inline test/test_sparse_csr.py:260]: Technically these indices don't have to be the same for each batch entry. A more powerful test would potentially modify them as well to be different for each batch entry.

**Review by cpuhrsch — COMMENTED** [review]
I think generally this looks fine, but let's wait until you resolved the current CI failures.

**Comment by IvanYashchuk** [comment]
@cpuhrsch, CI is all green now. Could you please take another look?

**Review by cpuhrsch — APPROVED** [review]

**Comment by malfet** [comment]
This broke internal builds with:
```
caffe2/aten/src/ATen/native/sparse/SparseCsrTensor.cpp:83:26: error: comparison of integers of different signs: 'int64_t' (aka 'long') and 'unsigned long' [-Werror,-Wsign-compare]
      crow_indices.dim() == size.size() - 1,
      ~~~~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~
```
Curious why it was not caught by OSS CI

**Comment by malfet** [comment]
> @malfet - presumably OSS CI doesn't throw an error for that particular warning.

We compile with `-Werror -Wall` using clang-9, so it should

**Comment by IvanYashchuk** [comment]
When I resolve the issue can this PR be merged again using bot commands or I must open a new PR?

**Comment by cpuhrsch** [comment]
@IvanYashchuk - we'll see. Let's try to merge it once it's ready.

**Comment by IvanYashchuk** [comment]
@cpuhrsch, could you please import this PR to see whether there are any internal build failures?

**Comment by cpuhrsch** [comment]
@IvanYashchuk - well then let's try rerunning those jobs again.

**Comment by cpuhrsch** [comment]
@IvanYashchuk - FYI there's a PR that aims to prevent a broken master lint job from holding up PRs that are built upon viable strict. https://github.com/pytorch/pytorch/pull/75199

**Comment by b0noI** [comment]
Internal errors:

Summary: 
stderr: caffe2/aten/src/ATen/native/sparse/cuda/SparseBlasImpl.cpp:983:20: error: lambda capture 'C_crow_indices' is not used [-Werror,-Wunused-lambda-capture]
  auto fix_nnz = [&C_crow_indices, &m](int nnz) -> int {
                  ~^~~~~~~~~~~~~~~
caffe2/aten/src/ATen/native/sparse/cuda/SparseBlasImpl.cpp:983:37: error: lambda capture 'm' is not used [-Werror,-Wunused-lambda-capture]
  auto fix_nnz = [&C_crow_indices, &m](int nnz) -> int {
                                 ~~~^
stderr: caffe2/aten/src/ATen/native/sparse/cuda/SparseBlasImpl.cpp:983:20: error: lambda capture 'C_crow_indices' is not used [-Werror,-Wunused-lambda-capture]
  auto fix_nnz = [&C_crow_indices, &m](int nnz) -> int {
                  ~^~~~~~~~~~~~~~~
caffe2/aten/src/ATen/native/sparse/cuda/SparseBlasImpl.cpp:983:37: error: lambda capture 'm' is not used [-Werror,-Wunused-lambda-capture]
  auto fix_nnz = [&C_crow_indices, &m](int nnz) -> int {
                                 ~~~^

**Comment by cpuhrsch** [comment]
@malfet - We might want to make this error part of the CI too

**Comment by cpuhrsch** [comment]
@IvanYashchuk - I merged master and added a simple fix for this and will attempt to merge again once the CI runs green
```
diff --git a/aten/src/ATen/native/sparse/cuda/SparseBlasImpl.cpp b/aten/src/ATen/native/sparse/cuda/SparseBlasImpl.cpp
index 27432431f7..7cfe1248fb 100644
--- a/aten/src/ATen/native/sparse/cuda/SparseBlasImpl.cpp
+++ b/aten/src/ATen/native/sparse/cuda/SparseBlasImpl.cpp
@@ -983,15 +983,21 @@ void add_out_sparse_csr(
   auto C_col_indices_ptr = C_col_indices.data_ptr<int>();

   // Windows compilers don't support nested macros
-  // so we need this lambda outside of the AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES
-  auto fix_nnz = [&C_crow_indices, &m](int nnz) -> int {
-    // For some reason POINTER_MODE_HOST is not working here
-    // Let's extract manually the nnz from the C_crow_indices
-    #if AT_ROCM_ENABLED()
+  // so we need this lambda outside of the
+  // AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES
+  auto fix_nnz = [
+#if AT_ROCM_ENABLED()
+                     &C_crow_indices,
+                     &m
+#endif
+  ](int nnz) -> int {
+// For some reason POINTER_MODE_HOST is not working here
+// Let's extract manually the nnz from the C_crow_indices
+#if AT_ROCM_ENABLED()
     return std::max({nnz, C_crow_indices.narrow(-1, m, 1).item<int>()});
-    #else
+#else
     return nnz;
-    #endif
+#endif
   };
```

**Comment by malfet** [comment]
> @malfet - We might want to make this error part of the CI too

We should have it already, but we do not have a single CUDA build with -Werror in OSS CI

**Comment by IvanYashchuk** [comment]
@cpuhrsch, let's try one more time? 🤞

**Pull request closed without merging**